core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexAliasesService.java

```diff
@@ -56,109 +62,114 @@
-                        IndexMetaData.Builder indexMetaDataBuilder = IndexMetaData.builder(indexMetaData);
-                        if (aliasAction.actionType() == AliasAction.Type.ADD) {
-                            String filter = aliasAction.filter();
-                            if (Strings.hasLength(filter)) {
-                                // parse the filter, in order to validate it
-                                IndexService indexService = indices.get(indexMetaData.getIndex());
-                                if (indexService == null) {
-                                    indexService = indicesService.indexService(indexMetaData.getIndex());
-                                    if (indexService == null) {
-                                        // temporarily create the index and add mappings so we can parse the filter
-                                        try {
-                                            indexService = indicesService.createIndex(nodeServicesProvider, indexMetaData, Collections.emptyList());
-                                            for (ObjectCursor<MappingMetaData> cursor : indexMetaData.getMappings().values()) {
-                                                MappingMetaData mappingMetaData = cursor.value;
-                                                indexService.mapperService().merge(mappingMetaData.type(), mappingMetaData.source(), MapperService.MergeReason.MAPPING_RECOVERY, false);
-                                            }
-                                        } catch (Exception e) {
-                                            logger.warn("[{}] failed to temporary create in order to apply alias action", e, indexMetaData.getIndex());
-                                            continue;
-                                        }
-                                        indicesToClose.add(indexMetaData.getIndex());
-                                    }
-                                    indices.put(indexMetaData.getIndex().getName(), indexService);
+    ClusterState innerExecute(ClusterState currentState, Iterable<AliasAction> actions) {
+        List<Index> indicesToClose = new ArrayList<>();
+        Map<String, IndexService> indices = new HashMap<>();
+        try {
+            boolean changed = false;
+            // Gather all the indexes that must be removed first so:
+            // 1. We don't cause error when attempting to replace an index with a alias of the same name.
+            // 2. We don't allow removal of aliases from indexes that we're just going to delete anyway. That'd be silly.
+            List<Index> indicesToDelete = new ArrayList<>();
+            for (AliasAction action : actions) {
+                if (action.removeIndex()) {
+                    IndexMetaData index = currentState.metaData().getIndices().get(action.getIndex());
+                    if (index == null) {
+                        throw new IndexNotFoundException(action.getIndex());
```

Comment:
You made me consider it but I still like it this way because I like the idea of explicitly telling the user that the thing they wanted to do couldn't be done. They can work it out from there. If we were exposing this operation over an HTTP DELETE then I'd want to pass a boolean down to allow ignoring unavailable I think. Because HTTP DELETE has to be idempotent and reply `200 OK` if the thing is already gone.